pallets/amm/src/tests.rs

```diff
@@ -0,0 +1,73 @@
+use super::*;
+use crate::{mock::*, Error};
+use frame_support::{assert_noop, assert_ok};
+
+#[test]
+fn add_liquidity_should_work() {
+    new_test_ext().execute_with(|| {
+        assert_ok!(AMM::add_liquidity(
+            Origin::signed(1.into()),
+            (DOT, XDOT),
+            (10, 20)
+        ));
+
+        assert_eq!(AMM::pools(XDOT, DOT).base_amount, 10);
+
+        assert_eq!(
+            AMM::liquidity_providers((AccountId(1u64), XDOT, DOT)).base_amount,
+            10
+        );
+
+        // Check balance is correct
+        assert_eq!(
+            <Test as Config>::Currency::free_balance(CurrencyId::DOT, &1.into()),
+            80
+        );
+        assert_eq!(
+            <Test as Config>::Currency::free_balance(CurrencyId::xDOT, &1.into()),
+            90
+        );
+    })
+}
+
+#[test]
+fn add_liquidity_should_not_work() {
+    new_test_ext().execute_with(|| {
+        let _ = AMM::add_liquidity(Origin::signed(1.into()), (DOT, XDOT), (10, 20));
+
+        assert_noop!(
+            AMM::add_liquidity(Origin::signed(1.into()), (DOT, XDOT), (10, 20)),
+            Error::<Test>::PoolAlreadyExists,
+        );
+    })
+}
+
+#[test]
+fn remove_liquidity_should_work() {
```

Comment:
This test should account for mores cases

1. A pool with a single LP provider who deposit tokens and withdraws their whole share (most simple case)
2. A pool with a single LP provider who deposit tokens and withdraws a portion of their total shares (simple case)
3. A pool with multiple LP providers who deposit tokens with varying ratios and some withdraw their shares (complex example)